src/Interpreters/ProcessList.h

```diff
@@ -145,6 +154,18 @@ class QueryStatus : public WithContext
 
     ~QueryStatus();
 
+    void allowRemoval() { kill_query_flags = KillQueryFlags::NONE; }
+    void forbidRemoval() { kill_query_flags = KillQueryFlags::FORBID_REMOVAL; }
```

Comment:
Do we need some assertion here as well? What transitions between states are possible and when? (I tried to draw state machine on piece of paper, but failed)